Refactor RenderLayer::scrollRectToVisible https://bugs.webkit.org/show_bug.cgi?id=236220

mrobinson opened the pull request:
#### ba34b343955eddf821d30bb4c681300f93d0fa38
<pre>
Refactor RenderLayer::scrollRectToVisible <a href="https://bugs.webkit.org/show_bug.cgi?id=236220">https://bugs.webkit.org/show_bug.cgi?id=236220</a>

Reviewed by Simon Fraser.

Source/WebCore:

No new tests. This should not change behavior.

Move RenderLayer::scrollRectToVisible to a static method in FrameView and combine it with
RenderObject::scrollRectToVisible. This also splits out the three code paths in this method
into three separate methods, two of them on FrameView and one of them on
RenderLayerScrollableArea.

In additions, various methods are moved and renamed to make them more
understandable. ScrollAlignment must move to platform, because getRectToExpose moves to
ScrollableArea and is renamed to getRectToExposeForScrollIntoView.

* Headers.cmake:
* PlatformMac.cmake:
* Sources.txt:
* WebCore.xcodeproj/project.pbxproj:
* accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::scrollToMakeVisible const):
* accessibility/atspi/AccessibilityObjectTextAtspi.cpp:
(WebCore::AccessibilityObjectAtspi::scrollToMakeVisible const):
* dom/Element.cpp:
(WebCore::Element::scrollIntoView):
(WebCore::Element::scrollIntoViewIfNeeded):
(WebCore::Element::scrollIntoViewIfNotVisible):
* editing/FrameSelection.cpp:
(WebCore::FrameSelection::revealSelection):
* page/FrameView.cpp:
(WebCore::FrameView::scrollToFocusedElementInternal):
(WebCore::FrameView::scrollRectToVisible):
(WebCore::scrollPositionChangeOptionsForElement):
(WebCore::FrameView::scrollRectToVisibleInChildView):
(WebCore::FrameView::scrollRectToVisibleInTopLevelView):
(WebCore::FrameView::scrollToAnchor):
(WebCore::FrameView::getPossiblyFixedRectToExpose const):
* page/FrameView.h:
* platform/ScrollAlignment.cpp: Renamed from Source/WebCore/rendering/ScrollAlignment.cpp.
* platform/ScrollAlignment.h: Renamed from Source/WebCore/rendering/ScrollAlignment.h.
* platform/ScrollableArea.cpp:
(WebCore::ScrollableArea::getRectToExposeForScrollIntoView const):
* platform/ScrollableArea.h:
* rendering/RenderLayer.cpp:
(WebCore::RenderLayer::shouldTryToScrollForScrollIntoView const):
(WebCore::RenderLayer::autoscroll):
(WebCore::expandScrollRectToVisibleTargetRectToIncludeScrollPadding): Deleted.
(WebCore::frameElementAndViewPermitScroll): Deleted.
(WebCore::RenderLayer::allowsCurrentScroll const): Deleted.
(WebCore::RenderLayer::scrollRectToVisible): Deleted.
(WebCore::RenderLayer::getRectToExpose const): Deleted.
* rendering/RenderLayer.h:
* rendering/RenderLayerScrollableArea.cpp:
(WebCore::RenderLayerScrollableArea::scrollRectToVisible):
(WebCore::RenderLayerScrollableArea::updateScrollPositionForScrollIntoView):
(WebCore::RenderLayerScrollableArea::updateScrollPosition): Deleted.
* rendering/RenderLayerScrollableArea.h:
* rendering/RenderObject.cpp:
(WebCore::RenderObject::scrollRectToVisible): Deleted.
* rendering/RenderObject.h:

Source/WebKitLegacy/mac:

* WebView/WebFrame.mm:
(-[WebFrame _scrollDOMRangeToVisible:]):
(-[WebFrame _scrollDOMRangeToVisible:withInset:]):

Canonical link: <a href="https://commits.webkit.org/252595@main">https://commits.webkit.org/252595@main</a>
</pre>

smfr commented on Source/WebCore/page/FrameView.cpp:
```diff
@@ -2483,7 +2484,144 @@ void FrameView::scrollToFocusedElementInternal()
+    if (options.revealMode == SelectionRevealMode::DoNotReveal)
+        return false;
+
+    RenderLayer* layer = renderer.enclosingLayer();
```
auto

smfr commented on Source/WebCore/page/FrameView.cpp:
```diff
@@ -2483,7 +2484,144 @@ void FrameView::scrollToFocusedElementInternal()
+        return false;
+
+    // FIXME: It would be nice to use RenderLayer::enclosingScrollableLayer here, but that seems to skip overflow:hidden layers.
+    LayoutRect adjustedRect = absoluteRect;
```
auto

smfr commented on Source/WebCore/page/FrameView.cpp:
```diff
@@ -2483,7 +2484,144 @@ void FrameView::scrollToFocusedElementInternal()
+void FrameView::scrollRectToVisibleInChildView(const LayoutRect& absoluteRect, bool insideFixed, const ScrollRectToVisibleOptions& options, const HTMLFrameOwnerElement* ownerElement)
+{
+    const HTMLFrameElementBase* frameElementBase = nullptr;
+    if (is<HTMLFrameElementBase>(*ownerElement))
```
dynamic_downcast<> would shrink this to one line.

smfr reviewed (approved):
This is a great change!

mrobinson committed:
Refactor RenderLayer::scrollRectToVisible https://bugs.webkit.org/show_bug.cgi?id=236220

Reviewed by Simon Fraser.

Source/WebCore:

No new tests. This should not change behavior.

Move RenderLayer::scrollRectToVisible to a static method in FrameView and combine it with
RenderObject::scrollRectToVisible. This also splits out the three code paths in this method
into three separate methods, two of them on FrameView and one of them on
RenderLayerScrollableArea.

In additions, various methods are moved and renamed to make them more
understandable. ScrollAlignment must move to platform, because getRectToExpose moves to
ScrollableArea and is renamed to getRectToExposeForScrollIntoView.

* Headers.cmake:
* PlatformMac.cmake:
* Sources.txt:
* WebCore.xcodeproj/project.pbxproj:
* accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::scrollToMakeVisible const):
* accessibility/atspi/AccessibilityObjectTextAtspi.cpp:
(WebCore::AccessibilityObjectAtspi::scrollToMakeVisible const):
* dom/Element.cpp:
(WebCore::Element::scrollIntoView):
(WebCore::Element::scrollIntoViewIfNeeded):
(WebCore::Element::scrollIntoViewIfNotVisible):
* editing/FrameSelection.cpp:
(WebCore::FrameSelection::revealSelection):
* page/FrameView.cpp:
(WebCore::FrameView::scrollToFocusedElementInternal):
(WebCore::FrameView::scrollRectToVisible):
(WebCore::scrollPositionChangeOptionsForElement):
(WebCore::FrameView::scrollRectToVisibleInChildView):
(WebCore::FrameView::scrollRectToVisibleInTopLevelView):
(WebCore::FrameView::scrollToAnchor):
(WebCore::FrameView::getPossiblyFixedRectToExpose const):
* page/FrameView.h:
* platform/ScrollAlignment.cpp: Renamed from Source/WebCore/rendering/ScrollAlignment.cpp.
* platform/ScrollAlignment.h: Renamed from Source/WebCore/rendering/ScrollAlignment.h.
* platform/ScrollableArea.cpp:
(WebCore::ScrollableArea::getRectToExposeForScrollIntoView const):
* platform/ScrollableArea.h:
* rendering/RenderLayer.cpp:
(WebCore::RenderLayer::shouldTryToScrollForScrollIntoView const):
(WebCore::RenderLayer::autoscroll):
(WebCore::expandScrollRectToVisibleTargetRectToIncludeScrollPadding): Deleted.
(WebCore::frameElementAndViewPermitScroll): Deleted.
(WebCore::RenderLayer::allowsCurrentScroll const): Deleted.
(WebCore::RenderLayer::scrollRectToVisible): Deleted.
(WebCore::RenderLayer::getRectToExpose const): Deleted.
* rendering/RenderLayer.h:
* rendering/RenderLayerScrollableArea.cpp:
(WebCore::RenderLayerScrollableArea::scrollRectToVisible):
(WebCore::RenderLayerScrollableArea::updateScrollPositionForScrollIntoView):
(WebCore::RenderLayerScrollableArea::updateScrollPosition): Deleted.
* rendering/RenderLayerScrollableArea.h:
* rendering/RenderObject.cpp:
(WebCore::RenderObject::scrollRectToVisible): Deleted.
* rendering/RenderObject.h:

Source/WebKitLegacy/mac:

* WebView/WebFrame.mm:
(-[WebFrame _scrollDOMRangeToVisible:]):
(-[WebFrame _scrollDOMRangeToVisible:withInset:]):

Canonical link: https://commits.webkit.org/252595@main
